correctness fm102 -- format edit descriptor and array sections

---------------------------------- executing test fm102
./fm102.out > fm102.log

fatal Fortran runtime error(.//src/fm102.f:221): Data edit descriptor 'I' may not be used with a CHARACTER data item
/bin/sh: line 1:  2072 Aborted                 (core dumped) ./fm102.out > fm102.log
make: *** [fm102.run] Error 134

--- a/flang/include/flang/Lower/AbstractConverter.h
+++ b/flang/include/flang/Lower/AbstractConverter.h
@@ -14,14 +14,16 @@
 #define FORTRAN_LOWER_ABSTRACTCONVERTER_H
 
 #include "flang/Common/Fortran.h"
+#include "flang/Lower/Support/BoxValue.h"
+#include "flang/Lower/Utils.h"
 #include "mlir/IR/Module.h"
-#include "Utils.h"
 
 namespace Fortran {
 namespace common {
 template <typename>
 class Reference;
 }
+
 namespace evaluate {
 struct DataRef;
 template <typename>
@@ -66,25 +68,26 @@ public:
   virtual bool lookupLabelSet(SymbolRef sym, pft::LabelSet &labelSet) = 0;
 
   /// Get the code defined by a label
-  virtual Fortran::lower::pft::Evaluation* lookupLabel(pft::Label label) = 0;
+  virtual Fortran::lower::pft::Evaluation *lookupLabel(pft::Label label) = 0;
 
   //===--------------------------------------------------------------------===//
   // Expressions
   //===--------------------------------------------------------------------===//
 
   /// Generate the address of the location holding the expression, someExpr
-  virtual mlir::Value genExprAddr(const SomeExpr &,
-                                  mlir::Location *loc = nullptr) = 0;
+  virtual fir::ExtendedValue genExprAddr(const SomeExpr &,
+                                         mlir::Location *loc = nullptr) = 0;
   /// Generate the address of the location holding the expression, someExpr
-  mlir::Value genExprAddr(const SomeExpr *someExpr, mlir::Location loc) {
+  fir::ExtendedValue genExprAddr(const SomeExpr *someExpr, mlir::Location loc) {
     return genExprAddr(*someExpr, &loc);
   }
 
   /// Generate the computations of the expression to produce a value
-  virtual mlir::Value genExprValue(const SomeExpr &,
-                                   mlir::Location *loc = nullptr) = 0;
+  virtual fir::ExtendedValue genExprValue(const SomeExpr &,
+                                          mlir::Location *loc = nullptr) = 0;
   /// Generate the computations of the expression, someExpr, to produce a value
-  mlir::Value genExprValue(const SomeExpr *someExpr, mlir::Location loc) {
+  fir::ExtendedValue genExprValue(const SomeExpr *someExpr,
+                                  mlir::Location loc) {
     return genExprValue(*someExpr, &loc);
   }
 

--- a/flang/include/flang/Lower/CallInterface.h
+++ b/flang/include/flang/Lower/CallInterface.h
@@ -31,7 +31,6 @@
 #include "mlir/IR/Function.h"
 #include <memory>
 #include <optional>
-#include <string>
 
 namespace Fortran::semantics {
 class Symbol;

--- a/flang/include/flang/Lower/FIRBuilder.h
+++ b/flang/include/flang/Lower/FIRBuilder.h
@@ -26,6 +26,10 @@
 #include "llvm/ADT/DenseMap.h"
 #include "llvm/ADT/Optional.h"
 
+namespace fir {
+class ExtendedValue;
+}
+
 namespace Fortran::lower {
 
 class AbstractConverter;
@@ -142,21 +146,13 @@ public:
   // Linkage helpers (inline). The default linkage is external.
   //===--------------------------------------------------------------------===//
 
-  mlir::StringAttr createCommonLinkage() {
-    return getStringAttr("common");
-  }
+  mlir::StringAttr createCommonLinkage() { return getStringAttr("common"); }
 
-  mlir::StringAttr createInternalLinkage() {
-    return getStringAttr("internal");
-  }
+  mlir::StringAttr createInternalLinkage() { return getStringAttr("internal"); }
 
-  mlir::StringAttr createLinkOnceLinkage() {
-    return getStringAttr("linkonce");
-  }
+  mlir::StringAttr createLinkOnceLinkage() { return getStringAttr("linkonce"); }
 
-  mlir::StringAttr createWeakLinkage() {
-    return getStringAttr("weak");
-  }
+  mlir::StringAttr createWeakLinkage() { return getStringAttr("weak"); }
 
   /// Get a function by name. If the function exists in the current module, it
   /// is returned. Otherwise, a null FuncOp is returned.
@@ -211,6 +207,9 @@ public:
   mlir::Value convertToIndexType(mlir::Location loc, mlir::Value val) {
     return createConvert(loc, getIndexType(), val);
   }
+
+  /// Create one of the shape ops given an extended value.
+  mlir::Value createShape(mlir::Location loc, const fir::ExtendedValue &exv);
 
 private:
   const fir::KindMapping &kindMap;

--- a/flang/include/flang/Optimizer/Dialect/FIROps.td
+++ b/flang/include/flang/Optimizer/Dialect/FIROps.td
@@ -2181,27 +2181,27 @@ def fir_IfOp : region_Op<"if", [NoRegionArguments]> {
   let results = (outs Variadic<AnyType>:$results);
 
   let regions = (region
-    SizedRegion<1>:$whereRegion,
-    AnyRegion:$otherRegion
+    SizedRegion<1>:$thenRegion,
+    AnyRegion:$elseRegion
   );
 
   let skipDefaultBuilders = 1;
   let builders = [
     OpBuilder<"OpBuilder &builder, OperationState &result, "
-              "Value cond, bool withOtherRegion">,
+              "Value cond, bool withElseRegion">,
     OpBuilder<"OpBuilder &builder, OperationState &result, "
-              "TypeRange resultTypes, Value cond, bool withOtherRegion">
+              "TypeRange resultTypes, Value cond, bool withElseRegion">
   ];
 
   let extraClassDeclaration = [{
     mlir::OpBuilder getWhereBodyBuilder() {
-      assert(!whereRegion().empty() && "Unexpected empty 'where' region.");
-      mlir::Block &body = whereRegion().front();
+      assert(!thenRegion().empty() && "Unexpected empty 'where' region.");
+      mlir::Block &body = thenRegion().front();
       return mlir::OpBuilder(&body, std::prev(body.end()));
     }
     mlir::OpBuilder getOtherBodyBuilder() {
-      assert(!otherRegion().empty() && "Unexpected empty 'other' region.");
-      mlir::Block &body = otherRegion().front();
+      assert(!elseRegion().empty() && "Unexpected empty 'other' region.");
+      mlir::Block &body = elseRegion().front();
       return mlir::OpBuilder(&body, std::prev(body.end()));
     }
   }];

--- a/flang/lib/Lower/Bridge.cpp
+++ b/flang/lib/Lower/Bridge.cpp
@@ -347,13 +347,18 @@ public:
     return iter->second;
   }
 
-  mlir::Value genExprAddr(const Fortran::lower::SomeExpr &expr,
-                          mlir::Location *loc = nullptr) override final {
-    return createFIRAddr(loc ? *loc : toLocation(), &expr);
+  fir::ExtendedValue genExprAddr(const Fortran::lower::SomeExpr &expr,
+                                 mlir::Location *loc = nullptr) override final {
+    Fortran::lower::ExpressionContext context;
+    return createSomeExtendedAddress(loc ? *loc : toLocation(), *this, expr,
+                                     localSymbols, context);
   }
-  mlir::Value genExprValue(const Fortran::lower::SomeExpr &expr,
-                           mlir::Location *loc = nullptr) override final {
-    return createFIRExpr(loc ? *loc : toLocation(), &expr);
+  fir::ExtendedValue
+  genExprValue(const Fortran::lower::SomeExpr &expr,
+               mlir::Location *loc = nullptr) override final {
+    Fortran::lower::ExpressionContext context;
+    return createSomeExtendedExpression(loc ? *loc : toLocation(), *this, expr,
+                                        localSymbols, context);
   }
   Fortran::evaluate::FoldingContext &getFoldingContext() override final {
     return foldingContext;
@@ -466,11 +471,6 @@ private:
   // Helper member functions
   //===--------------------------------------------------------------------===//
 
-  mlir::Value createFIRAddr(mlir::Location loc,
-                            const Fortran::semantics::SomeExpr *expr) {
-    return createSomeAddress(loc, *this, *expr, localSymbols);
-  }
-
   mlir::Value createFIRExpr(mlir::Location loc,
                             const Fortran::semantics::SomeExpr *expr) {
     return createSomeExpression(loc, *this, *expr, localSymbols);
@@ -555,13 +555,14 @@ private:
   void genFIRConditionalBranch(const Fortran::parser::ScalarLogicalExpr &expr,
                                mlir::Block *trueTarget,
                                mlir::Block *falseTarget) {
-    mlir::Value cond = genExprValue(*Fortran::semantics::GetExpr(expr));
+    mlir::Value cond =
+        createFIRExpr(toLocation(), Fortran::semantics::GetExpr(expr));
     genFIRConditionalBranch(cond, trueTarget, falseTarget);
   }
   void genFIRConditionalBranch(const Fortran::parser::ScalarLogicalExpr &expr,
                                Fortran::lower::pft::Evaluation *trueTarget,
                                Fortran::lower::pft::Evaluation *falseTarget) {
-    mlir::Value cond = genExprValue(*Fortran::semantics::GetExpr(expr));
+    auto cond = createFIRExpr(toLocation(), Fortran::semantics::GetExpr(expr));
     genFIRConditionalBranch(cond, trueTarget->block, falseTarget->block);
   }
 
@@ -638,13 +639,22 @@ private:
   template <typename A>
   std::pair<mlir::OpBuilder::InsertPoint, fir::IfOp>
   genIfCondition(const A *stmt, bool withElse = true) {
-    auto cond = genExprValue(*Fortran::semantics::GetExpr(
-        std::get<Fortran::parser::ScalarLogicalExpr>(stmt->t)));
+    auto cond = createFIRExpr(
+        toLocation(),
+        Fortran::semantics::GetExpr(
+            std::get<Fortran::parser::ScalarLogicalExpr>(stmt->t)));
     auto bcc = builder->createConvert(toLocation(), builder->getI1Type(), cond);
     auto ifOp = builder->create<fir::IfOp>(toLocation(), bcc, withElse);
     auto insPt = builder->saveInsertionPoint();
-    builder->setInsertionPointToStart(&ifOp.whereRegion().front());
+    builder->setInsertionPointToStart(&ifOp.thenRegion().front());
     return {insPt, ifOp};
+  }
+
+  mlir::Value genFIRLoopIndex(const Fortran::parser::ScalarExpr &x,
+                              mlir::Type t) {
+    auto loc = toLocation();
+    mlir::Value v = createFIRExpr(loc, Fortran::semantics::GetExpr(x));
+    return builder->createConvert(loc, t, v);
   }
 
   mlir::FuncOp getFunc(llvm::StringRef name, mlir::FunctionType ty) {
@@ -660,8 +670,8 @@ private:
     auto &eval = getEval();
     setCurrentPosition(stmt.v.source);
     assert(stmt.typedCall && "Call was not analyzed");
-    Fortran::semantics::SomeExpr expr{*stmt.typedCall};
     // Call statement lowering shares code with function call lowering.
+    Fortran::semantics::SomeExpr expr{*stmt.typedCall};
     auto res = createFIRExpr(toLocation(), &expr);
     if (!res)
       return; // "Normal" subroutine call.
@@ -701,10 +711,11 @@ private:
 
   void genFIR(const Fortran::parser::ComputedGotoStmt &stmt) {
     auto &eval = getEval();
-    auto selectExpr = genExprValue(*Fortran::semantics::GetExpr(
-        std::get<Fortran::parser::ScalarIntExpr>(stmt.t)));
-    llvm::SmallVector<int64_t, 10> indexList;
-    llvm::SmallVector<mlir::Block *, 10> blockList;
+    auto selectExpr = createFIRExpr(
+        toLocation(), Fortran::semantics::GetExpr(
+                          std::get<Fortran::parser::ScalarIntExpr>(stmt.t)));
+    llvm::SmallVector<int64_t, 8> indexList;
+    llvm::SmallVector<mlir::Block *, 8> blockList;
     int64_t index = 0;
     for (auto &label : std::get<std::list<Fortran::parser::Label>>(stmt.t)) {
       indexList.push_back(++index);
@@ -717,8 +728,9 @@ private:
 
   void genFIR(const Fortran::parser::ArithmeticIfStmt &stmt) {
     auto &eval = getEval();
-    auto expr = genExprValue(
-        *Fortran::semantics::GetExpr(std::get<Fortran::parser::Expr>(stmt.t)));
+    auto expr = createFIRExpr(
+        toLocation(),
+        Fortran::semantics::GetExpr(std::get<Fortran::parser::Expr>(stmt.t)));
     auto exprType = expr.getType();
     auto loc = toLocation();
     if (exprType.isSignlessInteger()) {
@@ -947,7 +959,8 @@ private:
                                            : genType(info.loopVariableSym);
     auto genControlValue = [&](const Fortran::semantics::SomeExpr *expr) {
       if (expr)
-        return builder->createConvert(loc, controlType, genExprValue(*expr));
+        return builder->createConvert(loc, controlType,
+                                      createFIRExpr(loc, expr));
       if (!info.hasRealControl)
         return builder->createIntegerConstant(loc, controlType, 1); // step
       auto one =
@@ -982,8 +995,8 @@ private:
       builder->create<fir::StoreOp>(loc, value, info.loopVariable);
       if (info.maskExpr) {
         auto ifOp = builder->create<fir::IfOp>(
-            loc, genExprValue(*info.maskExpr), /*withOtherRegion=*/false);
-        builder->setInsertionPointToStart(&ifOp.whereRegion().front());
+            loc, createFIRExpr(loc, info.maskExpr), /*withElseRegion=*/false);
+        builder->setInsertionPointToStart(&ifOp.thenRegion().front());
       }
       genLocalInitAssignments();
       return;
@@ -1025,7 +1038,7 @@ private:
       startBlock(info.maskBlock);
       auto latchBlock = getEval().getLastNestedEvaluation().block;
       assert(latchBlock && "missing masked concurrent loop latch block");
-      genFIRConditionalBranch(genExprValue(*info.maskExpr), info.bodyBlock,
+      genFIRConditionalBranch(createFIRExpr(loc, info.maskExpr), info.bodyBlock,
                               latchBlock);
     } else {
       genFIRConditionalBranch(cond, info.bodyBlock, info.exitBlock);
@@ -1093,11 +1106,11 @@ private:
           std::tie(insPt, nestedIf) = genIfCondition(s);
         } else if (auto *s = e.getIf<Fortran::parser::ElseIfStmt>()) {
           // otherwise block, then nested fir.if
-          builder->setInsertionPointToStart(&nestedIf.otherRegion().front());
+          builder->setInsertionPointToStart(&nestedIf.elseRegion().front());
           std::tie(std::ignore, nestedIf) = genIfCondition(s);
         } else if (e.isA<Fortran::parser::ElseStmt>()) {
           // otherwise block
-          builder->setInsertionPointToStart(&nestedIf.otherRegion().front());
+          builder->setInsertionPointToStart(&nestedIf.elseRegion().front());
         } else if (e.isA<Fortran::parser::EndIfStmt>()) {
           builder->restoreInsertionPoint(insPt);
         } else {
@@ -1204,8 +1217,9 @@ private:
     using ScalarExpr = Fortran::parser::Scalar<Fortran::parser::Expr>;
     MLIRContext *context = builder->getContext();
     auto loc = toLocation();
-    auto selectExpr = genExprValue(
-        *Fortran::semantics::GetExpr(std::get<ScalarExpr>(stmt.t)));
+    auto selectExpr = createFIRExpr(
+        toLocation(),
+        Fortran::semantics::GetExpr(std::get<ScalarExpr>(stmt.t)));
     auto selectType = selectExpr.getType();
     Fortran::lower::CharacterExprHelper helper{*builder, loc};
     if (helper.isCharacter(selectExpr.getType())) {
@@ -1221,7 +1235,8 @@ private:
       const auto v = Fortran::evaluate::ToInt64(*expr);
       valueList.push_back(
           v ? builder->createIntegerConstant(loc, selectType, *v)
-            : builder->createConvert(loc, selectType, genExprValue(*expr)));
+            : builder->createConvert(loc, selectType,
+                                     createFIRExpr(toLocation(), expr)));
     };
     for (Fortran::lower::pft::Evaluation *e = eval.controlSuccessor; e;
          e = e->controlSuccessor) {
@@ -1467,9 +1482,8 @@ private:
 
     for (auto s : shape) {
       if (s.has_value()) {
-        auto ub = builder->createConvert(
-            loc, idxTy,
-            genExprValue(Fortran::evaluate::AsGenericExpr(std::move(*s))));
+        auto e = Fortran::evaluate::AsGenericExpr(std::move(*s));
+        auto ub = builder->createConvert(loc, idxTy, createFIRExpr(loc, &e));
         auto up = builder->create<mlir::SubIOp>(loc, ub, one);
         extents.push_back(up);
       } else {
@@ -1586,9 +1600,9 @@ private:
                 // Conversions should have been inserted by semantic analysis,
                 // but they can be incorrect between the rhs and lhs. Correct
                 // that here.
-                mlir::Value addr = isPointer ? genExprValue(assign.lhs)
-                                             : genExprAddr(assign.lhs);
-                auto val = genExprValue(assign.rhs);
+                auto addr = fir::getBase(isPointer ? genExprValue(assign.lhs)
+                                                   : genExprAddr(assign.lhs));
+                auto val = createFIRExpr(loc, &assign.rhs);
                 // A function with multiple entry points returning different
                 // types tags all result variables with one of the largest
                 // types to allow them to share the same storage.  Assignment
@@ -1608,7 +1622,7 @@ private:
                 // Fortran 2018 10.2.1.3 p10 and p11
                 // Generating value for lhs to get fir.boxchar.
                 Fortran::lower::ExpressionContext context;
-                auto lhs = genExprAddr(assign.lhs);
+                auto lhs = fir::getBase(genExprAddr(assign.lhs));
                 auto rhs = createSomeExtendedExpression(
                     toLocation(), *this, assign.rhs, localSymbols, context);
                 Fortran::lower::CharacterExprHelper{*builder, loc}.createAssign(
@@ -1728,10 +1742,10 @@ private:
       // to the compiler-generated result variable.
       const auto &symbol = funit->getSubprogramSymbol();
       if (Fortran::semantics::HasAlternateReturns(symbol)) {
-        auto expr = Fortran::semantics::GetExpr(*stmt.v);
+        const auto *expr = Fortran::semantics::GetExpr(*stmt.v);
         assert(expr && "missing alternate return expression");
         auto altReturnIndex = builder->createConvert(
-            loc, builder->getIndexType(), genExprValue(*expr));
+            loc, builder->getIndexType(), createFIRExpr(loc, expr));
         builder->create<fir::StoreOp>(loc, altReturnIndex,
                                       getAltReturnResult(symbol));
       }
@@ -2077,7 +2091,7 @@ private:
           addr = charHelp.createEmboxChar(boxAddr, len);
         } else if (auto e = sba.getCharLenExpr()) {
           // Set/override LEN with an expression
-          len = genExprValue(*e);
+          len = createFIRExpr(loc, &*e);
           addr = charHelp.createEmboxChar(boxAddr, len);
         } else {
           // LEN is from the boxchar
@@ -2094,7 +2108,7 @@ private:
         if (auto c = sba.getCharLenConst())
           len = builder->createIntegerConstant(loc, idxTy, *c);
         else if (auto e = sba.getCharLenExpr())
-          len = genExprValue(*e);
+          len = createFIRExpr(loc, &*e);
         else
           len = builder->createIntegerConstant(loc, idxTy, sym.size());
         assert(!addr);
@@ -2168,8 +2182,10 @@ private:
         auto high = spec->ubound().GetExplicit();
         if (low && high) {
           // let the folder deal with the common `ub - 1 + 1` case
-          auto lb = genExprValue(Fortran::semantics::SomeExpr{*low});
-          auto ub = genExprValue(Fortran::semantics::SomeExpr{*high});
+          Fortran::semantics::SomeExpr lowEx{*low};
+          auto lb = createFIRExpr(loc, &lowEx);
+          Fortran::semantics::SomeExpr highEx{*high};
+          auto ub = createFIRExpr(loc, &highEx);
           auto ty = ub.getType();
           auto diff = builder->create<mlir::SubIOp>(loc, ty, ub, lb);
           auto one = builder->createIntegerConstant(loc, ty, 1);
@@ -2181,7 +2197,8 @@ private:
         }
         if (low && spec->ubound().isAssumed()) {
           // An assumed size array. The extent is not computed.
-          auto lb = genExprValue(Fortran::semantics::SomeExpr{*low});
+          Fortran::semantics::SomeExpr lowEx{*low};
+          auto lb = createFIRExpr(loc, &lowEx);
           lbounds.emplace_back(lb);
           extents.emplace_back(mlir::Value{});
         }
@@ -2358,9 +2375,8 @@ private:
             builder.create<fir::HasValueOp>(loc, cb);
           };
           // create the global object
-          global =
-              builder->createGlobal(loc, commonTy, commonName,
-                                    /*isConstant=*/false, initFunc);
+          global = builder->createGlobal(loc, commonTy, commonName,
+                                         /*isConstant=*/false, initFunc);
         }
         // introduce a local AddrOf and add it to the map
         auto addrOf = builder->create<fir::AddrOfOp>(loc, global.resultType(),

--- a/flang/lib/Lower/CallInterface.cpp
+++ b/flang/lib/Lower/CallInterface.cpp
@@ -144,7 +144,7 @@ mlir::Value Fortran::lower::CallerInterface::getResultLength() {
   assert(typeAndShape && "no result type");
   auto expr = AsGenericExpr(typeAndShape->LEN().value());
   if (Fortran::evaluate::IsConstantExpr(expr))
-    return converter.genExprValue(expr);
+    return fir::getBase(converter.genExprValue(expr));
   llvm_unreachable(
       "non constant result length on caller side not yet safely handled");
 }

--- a/flang/lib/Lower/FIRBuilder.cpp
+++ b/flang/lib/Lower/FIRBuilder.cpp
@@ -11,6 +11,7 @@
 #include "flang/Lower/Bridge.h"
 #include "flang/Lower/ComplexExpr.h"
 #include "flang/Lower/ConvertType.h"
+#include "flang/Lower/Support/BoxValue.h"
 #include "flang/Optimizer/Dialect/FIROpsSupport.h"
 #include "flang/Semantics/symbol.h"
 #include "llvm/Support/ErrorHandling.h"
@@ -186,4 +187,36 @@ fir::StringLitOp Fortran::lower::FirOpBuilder::createStringLit(
       fir::SequenceType::get(fir::SequenceType::Shape(1, data.size()), eleTy);
   return create<fir::StringLitOp>(loc, llvm::ArrayRef<mlir::Type>{arrTy},
                                   llvm::None, attrs);
+}
+
+mlir::Value
+Fortran::lower::FirOpBuilder::createShape(mlir::Location loc,
+                                          const fir::ExtendedValue &exv) {
+  auto ctor = [&](const fir::AbstractArrayBox &box) -> mlir::Value {
+    if (box.lboundsAllOne()) {
+      // Create a ShapeOp with nominal origin of all ones.
+      auto shapeTy = fir::ShapeType::get(getContext(), box.getExtents().size());
+      return create<fir::ShapeOp>(loc, shapeTy, box.getExtents());
+    }
+    // Create a ShapeShiftOp, as origin may not be all ones.
+    auto idxTy = getIndexType();
+    auto shapeTy =
+        fir::ShapeShiftType::get(getContext(), box.getExtents().size());
+    llvm::SmallVector<mlir::Value, 8> pairs;
+    for (const auto &pair : llvm::zip(box.getLBounds(), box.getExtents())) {
+      auto lb = createConvert(loc, idxTy, std::get<0>(pair));
+      pairs.push_back(lb);
+      auto ext = createConvert(loc, idxTy, std::get<1>(pair));
+      pairs.push_back(ext);
+    }
+    return create<fir::ShapeShiftOp>(loc, shapeTy, pairs);
+  };
+  return exv.match([&](const fir::ArrayBoxValue &box) { return ctor(box); },
+                   [&](const fir::CharArrayBoxValue &box) { return ctor(box); },
+                   [&](const fir::BoxValue &box) { return ctor(box); },
+                   [&](const auto &) {
+                     exv.dump();
+                     mlir::emitError(loc, "expected shape on entity");
+                     return mlir::Value{};
+                   });
 }

--- a/flang/lib/Lower/IO.cpp
+++ b/flang/lib/Lower/IO.cpp
@@ -26,8 +26,6 @@
 #include "flang/Semantics/tools.h"
 #include "mlir/Dialect/StandardOps/IR/Ops.h"
 
-#define TODO() llvm_unreachable("not yet implemented")
-
 using namespace Fortran::runtime::io;
 
 #define mkIOKey(X) mkKey(IONAME(X))
@@ -145,7 +143,7 @@ static mlir::Value genEndIO(Fortran::lower::AbstractConverter &converter,
     auto getIoMsg = getIORuntimeFunc<mkIOKey(GetIoMsg)>(loc, builder);
     auto ioMsgVar =
         Fortran::lower::CharacterExprHelper{builder, loc}.createUnboxChar(
-            converter.genExprAddr(csi.ioMsgExpr, loc));
+            fir::getBase(converter.genExprAddr(csi.ioMsgExpr, loc)));
     llvm::SmallVector<mlir::Value, 3> args{
         cookie,
         builder.createConvert(loc, getIoMsg.getType().getInput(1),
@@ -158,7 +156,7 @@ static mlir::Value genEndIO(Fortran::lower::AbstractConverter &converter,
   llvm::SmallVector<mlir::Value, 1> endArgs{cookie};
   auto call = builder.create<fir::CallOp>(loc, endIoStatement, endArgs);
   if (csi.ioStatExpr) {
-    auto ioStatVar = converter.genExprAddr(csi.ioStatExpr, loc);
+    auto ioStatVar = fir::getBase(converter.genExprAddr(csi.ioStatExpr, loc));
     auto ioStatResult = builder.createConvert(
         loc, converter.genType(*csi.ioStatExpr), call.getResult(0));
     builder.create<fir::StoreOp>(loc, ioStatResult, ioStatVar);
@@ -179,16 +177,18 @@ static void makeNextConditionalOn(Fortran::lower::FirOpBuilder &builder,
     // Either I/O calls do not need to be checked, or the next I/O call is the
     // first potentially erroneous call.
     return;
+
   // A previous I/O call for a statement returned the bool `ok`.  If this call
   // is in a fir.iterate_while loop, the result must be propagated up to the
   // loop scope.  That is done in genIoLoop, but it is enabled here.
-  auto ifOp =
-      inIterWhileLoop
-          ? builder.create<fir::IfOp>(loc, builder.getI1Type(), ok, true)
-          : builder.create<fir::IfOp>(loc, ok, /*withOtherwise=*/false);
+  mlir::TypeRange resTys;
+  if (inIterWhileLoop)
+    resTys = builder.getI1Type();
+  auto ifOp = builder.create<fir::IfOp>(loc, resTys, ok,
+                                        /*withElseRegion=*/inIterWhileLoop);
   if (!insertPt.isSet())
     insertPt = builder.saveInsertionPoint();
-  builder.setInsertionPointToStart(&ifOp.whereRegion().front());
+  builder.setInsertionPointToStart(&ifOp.thenRegion().front());
 }
 
 template <typename D>
@@ -244,33 +244,30 @@ genOutputItemList(Fortran::lower::AbstractConverter &converter,
     auto loc = converter.genLocation(pExpr.source);
     makeNextConditionalOn(builder, loc, insertPt, checkResult, ok,
                           inIterWhileLoop);
-    auto itemValue =
+    auto itemBox =
         converter.genExprValue(Fortran::semantics::GetExpr(pExpr), loc);
-    auto itemType = itemValue.getType();
-    auto outputFunc = getOutputFunc(loc, builder, itemType);
+    auto itemValue = fir::getBase(itemBox);
+    auto itemTy = itemValue.getType();
+    auto outputFunc = getOutputFunc(loc, builder, itemTy);
     auto argType = outputFunc.getType().getInput(1);
     llvm::SmallVector<mlir::Value, 3> outputFuncArgs = {cookie};
     Fortran::lower::CharacterExprHelper helper{builder, loc};
-    if (helper.isCharacter(itemType)) {
+    if (helper.isCharacter(itemTy)) {
       auto dataLen = helper.materializeCharacter(itemValue);
       outputFuncArgs.push_back(builder.createConvert(
           loc, outputFunc.getType().getInput(1), dataLen.first));
       outputFuncArgs.push_back(builder.createConvert(
           loc, outputFunc.getType().getInput(2), dataLen.second));
-    } else if (fir::isa_complex(itemType)) {
+    } else if (fir::isa_complex(itemTy)) {
       auto parts = Fortran::lower::ComplexExprHelper{builder, loc}.extractParts(
           itemValue);
       outputFuncArgs.push_back(parts.first);
       outputFuncArgs.push_back(parts.second);
     } else {
       if (argType.isa<fir::BoxType>()) {
-        mlir::Value shape;
-        mlir::Value slice;
-        llvm::SmallVector<mlir::Value, 8> lenParams;
-        // TODO: add dynamic shape op
-        TODO();
-        itemValue = builder.create<fir::EmboxOp>(
-            loc, fir::BoxType::get(itemType), itemValue, shape, slice, lenParams);
+        mlir::Value shape = builder.createShape(loc, itemBox);
+        itemValue = builder.create<fir::EmboxOp>(loc, fir::BoxType::get(itemTy),
+                                                 itemValue, shape);
       }
       itemValue = builder.createConvert(loc, argType, itemValue);
       outputFuncArgs.push_back(itemValue);
@@ -327,34 +324,34 @@ static void genInputItemList(Fortran::lower::AbstractConverter &converter,
     auto loc = converter.genLocation(pVar.GetSource());
     makeNextConditionalOn(builder, loc, insertPt, checkResult, ok,
                           inIterWhileLoop);
-    auto itemAddr =
+    auto itemBox =
         converter.genExprAddr(Fortran::semantics::GetExpr(pVar), loc);
+    auto itemAddr = fir::getBase(itemBox);
     if (auto box = dyn_cast<fir::EmboxCharOp>(itemAddr.getDefiningOp()))
       itemAddr = box.memref();
-    auto itemType = itemAddr.getType().cast<fir::ReferenceType>().getEleTy();
-    auto inputFunc = getInputFunc(loc, builder, itemType);
+    auto itemTy = fir::dyn_cast_ptrEleTy(itemAddr.getType());
+    if (!itemTy)
+      mlir::emitError(loc, "internal: unhandled input item type ")
+          << itemAddr.getType();
+    auto inputFunc = getInputFunc(loc, builder, itemTy);
     auto argType = inputFunc.getType().getInput(1);
     auto originalItemAddr = itemAddr;
     if (argType.isa<fir::BoxType>()) {
-      mlir::Value shape;
-      mlir::Value slice;
-      llvm::SmallVector<mlir::Value, 8> lenParams;
-      // TODO: add dynamic shape op
-      TODO();
-      itemAddr = builder.create<fir::EmboxOp>(loc, fir::BoxType::get(itemType),
-                                              itemAddr, shape, slice, lenParams);
+      auto shape = builder.createShape(loc, itemBox);
+      auto boxTy = fir::BoxType::get(itemTy);
+      itemAddr = builder.create<fir::EmboxOp>(loc, boxTy, itemAddr, shape);
     }
     itemAddr = builder.createConvert(loc, argType, itemAddr);
     llvm::SmallVector<mlir::Value, 8> inputFuncArgs = {cookie, itemAddr};
     Fortran::lower::CharacterExprHelper helper{builder, loc};
-    if (helper.isCharacter(itemType)) {
+    if (helper.isCharacter(itemTy)) {
       auto len = helper.materializeCharacter(originalItemAddr).second;
       inputFuncArgs.push_back(
           builder.createConvert(loc, inputFunc.getType().getInput(2), len));
-    } else if (itemType.isa<mlir::IntegerType>()) {
+    } else if (itemTy.isa<mlir::IntegerType>()) {
       inputFuncArgs.push_back(builder.create<mlir::ConstantOp>(
           loc, builder.getI32IntegerAttr(
-                   itemType.cast<mlir::IntegerType>().getWidth() / 8)));
+                   itemTy.cast<mlir::IntegerType>().getWidth() / 8)));
     }
     ok =
         builder.create<fir::CallOp>(loc, inputFunc, inputFuncArgs).getResult(0);
@@ -377,9 +374,9 @@ static void genIoLoop(Fortran::lower::AbstractConverter &converter,
   const auto &loopSym = *control.name.thing.thing.symbol;
   auto loopVar = converter.getSymbolAddress(loopSym);
   auto genControlValue = [&](const Fortran::parser::ScalarIntExpr &expr) {
-    return builder.createConvert(
-        loc, builder.getIndexType(),
+    auto v = fir::getBase(
         converter.genExprValue(*Fortran::semantics::GetExpr(expr)));
+    return builder.createConvert(loc, builder.getIndexType(), v);
   };
   auto lowerValue = genControlValue(control.lower);
   auto upperValue = genControlValue(control.upper);
@@ -424,10 +421,10 @@ static void genIoLoop(Fortran::lower::AbstractConverter &converter,
   for (auto *op = builder.getBlock()->getParentOp(); isa<fir::IfOp>(op);
        op = op->getBlock()->getParentOp()) {
     auto ifOp = dyn_cast<fir::IfOp>(op);
-    auto *lastOp = &ifOp.whereRegion().front().back();
+    auto *lastOp = &ifOp.thenRegion().front().back();
     builder.setInsertionPointAfter(lastOp);
     builder.create<fir::ResultOp>(loc, lastOp->getResult(0)); // runtime result
-    builder.setInsertionPointToStart(&ifOp.otherRegion().front());
+    builder.setInsertionPointToStart(&ifOp.elseRegion().front());
     builder.create<fir::ResultOp>(loc, falseValue); // known false result
   }
   builder.restoreInsertionPoint(insertPt);
@@ -487,7 +484,7 @@ lowerStringLit(Fortran::lower::AbstractConverter &converter, mlir::Location loc,
                mlir::Type ty2 = {}) {
   auto &builder = converter.getFirOpBuilder();
   auto *expr = Fortran::semantics::GetExpr(syntax);
-  auto str = converter.genExprValue(expr, loc);
+  auto str = fir::getBase(converter.genExprValue(expr, loc));
   Fortran::lower::CharacterExprHelper helper{builder, loc};
   auto dataLen = helper.materializeCharacterOrSequence(str);
   auto buff = builder.createConvert(loc, strTy, dataLen.first);
@@ -531,7 +528,8 @@ mlir::Value genIntIOOption(Fortran::lower::AbstractConverter &converter,
   auto &builder = converter.getFirOpBuilder();
   mlir::FuncOp ioFunc = getIORuntimeFunc<A>(loc, builder);
   mlir::FunctionType ioFuncTy = ioFunc.getType();
-  auto expr = converter.genExprValue(Fortran::semantics::GetExpr(spec.v), loc);
+  auto expr = fir::getBase(
+      converter.genExprValue(Fortran::semantics::GetExpr(spec.v), loc));
   auto val = builder.createConvert(loc, ioFuncTy.getInput(1), expr);
   llvm::SmallVector<mlir::Value, 4> ioArgs = {cookie, val};
   return builder.create<fir::CallOp>(loc, ioFunc, ioArgs).getResult(0);
@@ -906,22 +904,24 @@ constexpr bool isDataTransferInternal<Fortran::parser::PrintStmt>(
 /// If the variable `var` is an array or of a KIND other than the default
 /// (normally 1), then a descriptor is required by the runtime IO API. This
 /// condition holds even in F77 sources.
-static llvm::Optional<mlir::Value> getVariableBufferRequiredDescriptor(
+static llvm::Optional<fir::ExtendedValue> getVariableBufferRequiredDescriptor(
     Fortran::lower::AbstractConverter &converter,
     const Fortran::parser::Variable &var) {
-  auto varAddr = converter.genExprAddr(var.typedExpr->v.value());
+  auto varBox = converter.genExprAddr(var.typedExpr->v.value());
   auto defCharKind =
       fir::getKindMapping(converter.getModuleOp())->defaultCharacterKind();
+  auto varAddr = fir::getBase(varBox);
   if (Fortran::lower::CharacterExprHelper::getCharacterOrSequenceKind(
           varAddr.getType()) != defCharKind)
-    return varAddr;
+    return varBox;
   if (Fortran::lower::CharacterExprHelper::isArray(varAddr.getType()))
-    return varAddr;
+    return varBox;
   return llvm::None;
 }
 
 template <typename A>
-static llvm::Optional<mlir::Value> getIfDataTransferInternalRequiresDescriptor(
+static llvm::Optional<fir::ExtendedValue>
+getIfDataTransferInternalRequiresDescriptor(
     Fortran::lower::AbstractConverter &converter, const A &stmt) {
   if (stmt.iounit.has_value())
     if (auto *var = std::get_if<Fortran::parser::Variable>(&stmt.iounit->u))
@@ -932,7 +932,7 @@ static llvm::Optional<mlir::Value> getIfDataTransferInternalRequiresDescriptor(
   return llvm::None;
 }
 template <>
-inline llvm::Optional<mlir::Value>
+inline llvm::Optional<fir::ExtendedValue>
 getIfDataTransferInternalRequiresDescriptor<Fortran::parser::PrintStmt>(
     Fortran::lower::AbstractConverter &, const Fortran::parser::PrintStmt &) {
   return llvm::None;
@@ -1032,7 +1032,7 @@ lowerReferenceAsStringSelect(
   // Create the unit case which should result in an error.
   auto *unitBlock = block->splitBlock(builder.getInsertionPoint());
   builder.setInsertionPointToEnd(unitBlock);
-  
+
   // Crash the program.
   builder.create<fir::UnreachableOp>(loc);
 
@@ -1041,7 +1041,7 @@ lowerReferenceAsStringSelect(
 
   // Lower the selectOp
   builder.setInsertionPointToEnd(startBlock);
-  auto label = converter.genExprValue(&expr, loc);
+  auto label = fir::getBase(converter.genExprValue(&expr, loc));
   builder.create<fir::SelectOp>(loc, label, indexList, blockList);
 
   builder.setInsertionPointToEnd(endBlock);
@@ -1119,8 +1119,8 @@ genBuffer(Fortran::lower::AbstractConverter &converter, mlir::Location loc,
       *e, Fortran::common::TypeCategory::Character));
   // Helper to query [BUFFER, LEN].
   Fortran::lower::CharacterExprHelper helper(builder, loc);
-  auto dataLen =
-      helper.materializeCharacterOrSequence(converter.genExprValue(*e));
+  auto dataLen = helper.materializeCharacterOrSequence(
+      fir::getBase(converter.genExprValue(*e)));
   auto buff = builder.createConvert(loc, strTy, dataLen.first);
   auto len = builder.createConvert(loc, lenTy, dataLen.second);
   return {buff, len};
@@ -1137,12 +1137,27 @@ getBuffer(Fortran::lower::AbstractConverter &converter, mlir::Location loc,
 }
 
 /// Create a boxed value to be passed to the runtime. `descRef` is the reference
-/// to an entity that is to be boxed.  The box will include a shape (or
-/// shape_shift) argument if the entity is an array.
+/// to an entity that is to be boxed. Array entities are boxed with a shape.
 static mlir::Value getDescriptor(Fortran::lower::AbstractConverter &converter,
-                                 mlir::Location loc, mlir::Value descRef,
+                                 mlir::Location loc,
+                                 const fir::ExtendedValue &descRef,
                                  mlir::Type toType) {
-  TODO();
+  auto &builder = converter.getFirOpBuilder();
+  mlir::Value embox;
+  auto addr = fir::getBase(descRef);
+  auto eleTy = fir::dyn_cast_ptrEleTy(addr.getType());
+  if (!eleTy)
+    mlir::emitError(loc, "internal: expected a memory reference type ")
+        << addr.getType();
+  auto boxTy = fir::BoxType::get(eleTy);
+  mlir::Value shape;
+  if (descRef.match([](const fir::ArrayBoxValue &) { return true; },
+                    [](const fir::CharArrayBoxValue &) { return true; },
+                    [](const fir::BoxValue &) { return true; },
+                    [](const auto &) { return false; }))
+    shape = builder.createShape(loc, descRef);
+  embox = builder.create<fir::EmboxOp>(loc, boxTy, addr, shape);
+  return builder.createConvert(loc, toType, embox);
 }
 
 static mlir::Value genIOUnit(Fortran::lower::AbstractConverter &converter,
@@ -1151,7 +1166,8 @@ static mlir::Value genIOUnit(Fortran::lower::AbstractConverter &converter,
                              mlir::Type ty) {
   auto &builder = converter.getFirOpBuilder();
   if (auto *e = std::get_if<Fortran::parser::FileUnitNumber>(&iounit.u)) {
-    auto ex = converter.genExprValue(Fortran::semantics::GetExpr(*e), loc);
+    auto ex = fir::getBase(
+        converter.genExprValue(Fortran::semantics::GetExpr(*e), loc));
     return builder.createConvert(loc, ty, ex);
   }
   return builder.create<mlir::ConstantOp>(
@@ -1178,8 +1194,8 @@ static mlir::Value genBasicIOStmt(Fortran::lower::AbstractConverter &converter,
   auto loc = converter.getCurrentLocation();
   auto beginFunc = getIORuntimeFunc<K>(loc, builder);
   mlir::FunctionType beginFuncTy = beginFunc.getType();
-  auto unit = converter.genExprValue(
-      getExpr<Fortran::parser::FileUnitNumber>(stmt), loc);
+  auto unit = fir::getBase(converter.genExprValue(
+      getExpr<Fortran::parser::FileUnitNumber>(stmt), loc));
   auto un = builder.createConvert(loc, beginFuncTy.getInput(0), unit);
   auto file = locationToFilename(converter, loc, beginFuncTy.getInput(1));
   auto line = locationToLineNo(builder, loc, beginFuncTy.getInput(2));
@@ -1229,8 +1245,8 @@ Fortran::lower::genOpenStatement(Fortran::lower::AbstractConverter &converter,
   if (hasMem<Fortran::parser::FileUnitNumber>(stmt)) {
     beginFunc = getIORuntimeFunc<mkIOKey(BeginOpenUnit)>(loc, builder);
     mlir::FunctionType beginFuncTy = beginFunc.getType();
-    auto unit = converter.genExprValue(
-        getExpr<Fortran::parser::FileUnitNumber>(stmt), loc);
+    auto unit = fir::getBase(converter.genExprValue(
+        getExpr<Fortran::parser::FileUnitNumber>(stmt), loc));
     beginArgs.push_back(
         builder.createConvert(loc, beginFuncTy.getInput(0), unit));
     beginArgs.push_back(
@@ -1274,13 +1290,13 @@ Fortran::lower::genWaitStatement(Fortran::lower::AbstractConverter &converter,
       hasId ? getIORuntimeFunc<mkIOKey(BeginWait)>(loc, builder)
             : getIORuntimeFunc<mkIOKey(BeginWaitAll)>(loc, builder);
   mlir::FunctionType beginFuncTy = beginFunc.getType();
-  auto unit = converter.genExprValue(
-      getExpr<Fortran::parser::FileUnitNumber>(stmt), loc);
+  auto unit = fir::getBase(converter.genExprValue(
+      getExpr<Fortran::parser::FileUnitNumber>(stmt), loc));
   auto un = builder.createConvert(loc, beginFuncTy.getInput(0), unit);
   llvm::SmallVector<mlir::Value, 4> args{un};
   if (hasId) {
-    auto id =
-        converter.genExprValue(getExpr<Fortran::parser::IdExpr>(stmt), loc);
+    auto id = fir::getBase(
+        converter.genExprValue(getExpr<Fortran::parser::IdExpr>(stmt), loc));
     args.push_back(builder.createConvert(loc, beginFuncTy.getInput(1), id));
   }
   auto cookie = builder.create<fir::CallOp>(loc, beginFunc, args).getResult(0);
@@ -1379,7 +1395,7 @@ void genBeginCallArguments(llvm::SmallVector<mlir::Value, 8> &ioArgs,
                            mlir::FunctionType ioFuncTy, bool isFormatted,
                            bool isList, bool isIntern, bool isOtherIntern,
                            bool isAsynch, bool isNml,
-                           const llvm::Optional<mlir::Value> &descRef) {
+                           const llvm::Optional<fir::ExtendedValue> &descRef) {
   auto &builder = converter.getFirOpBuilder();
   if constexpr (hasIOCtrl) {
     // READ/WRITE cases have a wide variety of argument permutations
@@ -1475,7 +1491,7 @@ genDataTransferStmt(Fortran::lower::AbstractConverter &converter,
   const bool isFormatted = isDataTransferFormatted(stmt);
   const bool isList = isFormatted ? isDataTransferList(stmt) : false;
   const bool isIntern = isDataTransferInternal(stmt);
-  llvm::Optional<mlir::Value> descRef =
+  llvm::Optional<fir::ExtendedValue> descRef =
       isIntern ? getIfDataTransferInternalRequiresDescriptor(converter, stmt)
                : llvm::None;
   const bool isOtherIntern = descRef.hasValue();
@@ -1591,7 +1607,7 @@ mlir::Value genInquireSpec<Fortran::parser::InquireSpec::CharVar>(
   auto specFuncTy = specFunc.getType();
   const auto *varExpr = Fortran::semantics::GetExpr(
       std::get<Fortran::parser::ScalarDefaultCharVariable>(var.t));
-  auto str = converter.genExprValue(varExpr, loc);
+  auto str = fir::getBase(converter.genExprValue(varExpr, loc));
   Fortran::lower::CharacterExprHelper helper(builder, loc);
   auto data = helper.materializeCharacter(str);
   llvm::SmallVector<mlir::Value, 8> args = {
@@ -1622,8 +1638,11 @@ mlir::Value genInquireSpec<Fortran::parser::InquireSpec::IntVar>(
   auto specFuncTy = specFunc.getType();
   const auto *varExpr = Fortran::semantics::GetExpr(
       std::get<Fortran::parser::ScalarIntVariable>(var.t));
-  auto addr = converter.genExprAddr(varExpr, loc);
+  auto addr = fir::getBase(converter.genExprAddr(varExpr, loc));
   auto eleTy = fir::dyn_cast_ptrEleTy(addr.getType());
+  if (!eleTy)
+    mlir::emitError(loc, "internal: expected a memory reference type ")
+        << addr.getType();
   auto bitWidth = eleTy.cast<mlir::IntegerType>().getWidth();
   auto idxTy = builder.getIndexType();
   auto kind = builder.createIntegerConstant(loc, idxTy, bitWidth / 8);
@@ -1654,11 +1673,11 @@ mlir::Value genInquireSpec<Fortran::parser::InquireSpec::LogVar>(
       pendId ? getIORuntimeFunc<mkIOKey(InquirePendingId)>(loc, builder)
              : getIORuntimeFunc<mkIOKey(InquireLogical)>(loc, builder);
   auto specFuncTy = specFunc.getType();
-  auto addr = converter.genExprAddr(
+  auto addr = fir::getBase(converter.genExprAddr(
       Fortran::semantics::GetExpr(
           std::get<Fortran::parser::Scalar<
               Fortran::parser::Logical<Fortran::parser::Variable>>>(var.t)),
-      loc);
+      loc));
   llvm::SmallVector<mlir::Value, 8> args = {
       builder.createConvert(loc, specFuncTy.getInput(0), cookie)};
   if (pendId)
@@ -1682,8 +1701,8 @@ lowerIdExpr(Fortran::lower::AbstractConverter &converter, mlir::Location loc,
     if (mlir::Value v =
             std::visit(Fortran::common::visitors{
                            [&](const Fortran::parser::IdExpr &idExpr) {
-                             return converter.genExprValue(
-                                 Fortran::semantics::GetExpr(idExpr), loc);
+                             return fir::getBase(converter.genExprValue(
+                                 Fortran::semantics::GetExpr(idExpr), loc));
                            },
                            [](const auto &) { return mlir::Value{}; }},
                        spec.u))
@@ -1735,16 +1754,16 @@ mlir::Value Fortran::lower::genInquireStatement(
     // File unit call.
     beginFunc = getIORuntimeFunc<mkIOKey(BeginInquireUnit)>(loc, builder);
     mlir::FunctionType beginFuncTy = beginFunc.getType();
-    beginArgs = {
-        builder.createConvert(loc, beginFuncTy.getInput(0),
-                              converter.genExprValue(exprPair.first, loc)),
-        locationToFilename(converter, loc, beginFuncTy.getInput(1)),
-        locationToLineNo(builder, loc, beginFuncTy.getInput(2))};
+    beginArgs = {builder.createConvert(
+                     loc, beginFuncTy.getInput(0),
+                     fir::getBase(converter.genExprValue(exprPair.first, loc))),
+                 locationToFilename(converter, loc, beginFuncTy.getInput(1)),
+                 locationToLineNo(builder, loc, beginFuncTy.getInput(2))};
   } else if (inquireFileName()) {
     // Filename call.
     beginFunc = getIORuntimeFunc<mkIOKey(BeginInquireFile)>(loc, builder);
     mlir::FunctionType beginFuncTy = beginFunc.getType();
-    auto file = converter.genExprValue(exprPair.first, loc);
+    auto file = fir::getBase(converter.genExprValue(exprPair.first, loc));
     // Helper to query [BUFFER, LEN].
     Fortran::lower::CharacterExprHelper helper(builder, loc);
     auto dataLen = helper.materializeCharacter(file);

--- a/flang/lib/Lower/OpenACC.cpp
+++ b/flang/lib/Lower/OpenACC.cpp
@@ -89,8 +89,8 @@ static void genACC(Fortran::lower::AbstractConverter &converter,
           if (const auto &gangNumValue =
                   std::get<std::optional<Fortran::parser::ScalarIntExpr>>(
                       x.t)) {
-            gangNum = converter.genExprValue(
-                *Fortran::semantics::GetExpr(gangNumValue.value()));
+            gangNum = fir::getBase(converter.genExprValue(
+                *Fortran::semantics::GetExpr(gangNumValue.value())));
             operands.push_back(gangNum);
           }
           if (const auto &gangStaticValue =
@@ -99,8 +99,8 @@ static void genACC(Fortran::lower::AbstractConverter &converter,
                 std::get<std::optional<Fortran::parser::ScalarIntExpr>>(
                     gangStaticValue.value().t);
             if (expr) {
-              gangStatic = converter.genExprValue(
-                  *Fortran::semantics::GetExpr(*expr));
+              gangStatic = fir::getBase(
+                  converter.genExprValue(*Fortran::semantics::GetExpr(*expr)));
             } else {
               // * was passed as value and will be represented as a -1 constant
               // integer.
@@ -116,8 +116,8 @@ static void genACC(Fortran::lower::AbstractConverter &converter,
                      std::get_if<Fortran::parser::AccClause::Worker>(
                          &clause.u)) {
         if (workerClause->v) {
-          workerNum = converter.genExprValue(
-              *Fortran::semantics::GetExpr(*workerClause->v));
+          workerNum = fir::getBase(converter.genExprValue(
+              *Fortran::semantics::GetExpr(*workerClause->v)));
           operands.push_back(workerNum);
         }
         executionMapping |= mlir::acc::OpenACCExecMapping::WORKER;
@@ -125,8 +125,8 @@ static void genACC(Fortran::lower::AbstractConverter &converter,
                      std::get_if<Fortran::parser::AccClause::Vector>(
                          &clause.u)) {
         if (vectorClause->v) {
-          vectorLength = converter.genExprValue(
-              *Fortran::semantics::GetExpr(*vectorClause->v));
+          vectorLength = fir::getBase(converter.genExprValue(
+              *Fortran::semantics::GetExpr(*vectorClause->v)));
           operands.push_back(vectorLength);
         }
         executionMapping |= mlir::acc::OpenACCExecMapping::VECTOR;
@@ -139,8 +139,8 @@ static void genACC(Fortran::lower::AbstractConverter &converter,
                   accTileExpr.t);
           ++tileOperands;
           if (expr) {
-            operands.push_back(converter.genExprValue(
-                *Fortran::semantics::GetExpr(*expr)));
+            operands.push_back(fir::getBase(
+                converter.genExprValue(*Fortran::semantics::GetExpr(*expr))));
           } else {
             // * was passed as value and will be represented as a -1 constant
             // integer.

--- a/flang/lib/Lower/OpenMP.cpp
+++ b/flang/lib/Lower/OpenMP.cpp
@@ -99,8 +99,8 @@ genOMP(Fortran::lower::AbstractConverter &absConv,
       if (const auto &numThreadsClause =
               std::get_if<Fortran::parser::OmpClause::NumThreads>(&clause.u)) {
         // OMPIRBuilder expects `NUM_THREAD` clause as a `Value`.
-        numThreads = absConv.genExprValue(
-            *Fortran::semantics::GetExpr(numThreadsClause->v));
+        numThreads = fir::getBase(absConv.genExprValue(
+            *Fortran::semantics::GetExpr(numThreadsClause->v)));
       }
     }
     llvm::ArrayRef<mlir::Type> argTy;

--- a/flang/lib/Lower/Runtime.cpp
+++ b/flang/lib/Lower/Runtime.cpp
@@ -85,7 +85,7 @@ void Fortran::lower::genStopStatement(
           std::get<std::optional<Fortran::parser::StopCode>>(stmt.t)) {
     auto expr = Fortran::semantics::GetExpr(*code);
     assert(expr && "failed getting typed expression");
-    operands.push_back(converter.genExprValue(*expr));
+    operands.push_back(fir::getBase(converter.genExprValue(*expr)));
   } else {
     operands.push_back(
         builder.createIntegerConstant(loc, calleeType.getInput(0), 0));
@@ -101,7 +101,7 @@ void Fortran::lower::genStopStatement(
           std::get<std::optional<Fortran::parser::ScalarLogicalExpr>>(stmt.t)) {
     auto expr = Fortran::semantics::GetExpr(*quiet);
     assert(expr && "failed getting typed expression");
-    operands.push_back(converter.genExprValue(*expr));
+    operands.push_back(fir::getBase(converter.genExprValue(*expr)));
   } else {
     operands.push_back(
         builder.createIntegerConstant(loc, calleeType.getInput(2), 0));

--- a/flang/lib/Optimizer/Dialect/FIROps.cpp
+++ b/flang/lib/Optimizer/Dialect/FIROps.cpp
@@ -1402,7 +1402,7 @@ static mlir::ParseResult parseIfOp(OpAsmParser &parser,
 }
 
 static LogicalResult verify(fir::IfOp op) {
-  if (op.getNumResults() != 0 && op.otherRegion().empty())
+  if (op.getNumResults() != 0 && op.elseRegion().empty())
     return op.emitOpError("must have an else block if defining values");
 
   return mlir::success();
@@ -1415,11 +1415,11 @@ static void print(mlir::OpAsmPrinter &p, fir::IfOp op) {
     p << " -> (" << op.getResultTypes() << ')';
     printBlockTerminators = true;
   }
-  p.printRegion(op.whereRegion(), /*printEntryBlockArgs=*/false,
+  p.printRegion(op.thenRegion(), /*printEntryBlockArgs=*/false,
                 printBlockTerminators);
 
   // Print the 'else' regions if it exists and has a block.
-  auto &otherReg = op.otherRegion();
+  auto &otherReg = op.elseRegion();
   if (!otherReg.empty()) {
     p << " else";
     p.printRegion(otherReg, /*printEntryBlockArgs=*/false,

--- a/flang/lib/Optimizer/Transforms/AffinePromotion.cpp
+++ b/flang/lib/Optimizer/Transforms/AffinePromotion.cpp
@@ -507,7 +507,7 @@ public:
                   mlir::PatternRewriter &rewriter) const override {
     LLVM_DEBUG(llvm::dbgs() << "AffineIfConversion: rewriting if:\n";
                op.dump(););
-    auto &ifOps = op.whereRegion().front().getOperations();
+    auto &ifOps = op.thenRegion().front().getOperations();
     auto affineCondition = AffineIfCondition(op.condition());
     if (!affineCondition.integerSet) {
       LLVM_DEBUG(
@@ -517,12 +517,12 @@ public:
     }
     auto affineIf = rewriter.create<mlir::AffineIfOp>(
         op.getLoc(), affineCondition.integerSet.getValue(),
-        affineCondition.affineArgs, !op.otherRegion().empty());
+        affineCondition.affineArgs, !op.elseRegion().empty());
     rewriter.startRootUpdate(affineIf);
     affineIf.getThenBlock()->getOperations().splice(
         --affineIf.getThenBlock()->end(), ifOps, ifOps.begin(), --ifOps.end());
-    if (!op.otherRegion().empty()) {
-      auto &otherOps = op.otherRegion().front().getOperations();
+    if (!op.elseRegion().empty()) {
+      auto &otherOps = op.elseRegion().front().getOperations();
       affineIf.getElseBlock()->getOperations().splice(
           --affineIf.getElseBlock()->end(), otherOps, otherOps.begin(),
           --otherOps.end());

--- a/flang/lib/Optimizer/Transforms/RewriteLoop.cpp
+++ b/flang/lib/Optimizer/Transforms/RewriteLoop.cpp
@@ -157,7 +157,7 @@ public:
 
     // Move blocks from the "then" region to the region containing 'fir.if',
     // place it before the continuation block, and branch to it.
-    auto &ifOpRegion = ifOp.whereRegion();
+    auto &ifOpRegion = ifOp.thenRegion();
     auto *ifOpBlock = &ifOpRegion.front();
     auto *ifOpTerminator = ifOpRegion.back().getTerminator();
     auto ifOpTerminatorOperands = ifOpTerminator->getOperands();
@@ -170,7 +170,7 @@ public:
     // 'fir.if', place it before the continuation block and branch to it.  It
     // will be placed after the "then" regions.
     auto *otherwiseBlock = continueBlock;
-    auto &otherwiseRegion = ifOp.otherRegion();
+    auto &otherwiseRegion = ifOp.elseRegion();
     if (!otherwiseRegion.empty()) {
       otherwiseBlock = &otherwiseRegion.front();
       auto *otherwiseTerm = otherwiseRegion.back().getTerminator();

